### PR TITLE
Permission system, Preference API, mongoose v7, bug fixes

### PR DESCRIPTION
**Pull request by emilianya**

### CODE REVIEW

**Review by emilianya — COMMENTED** [review]
LGTM!!!

**Comment by emilianya** [comment]
(too bad i cant review myself)

**Review by byemc — APPROVED** [review]
hi i don't work here but these changes don't break anything

i am an unsuspecting user im sure theres no breaking changes

**Review by nyakase — APPROVED** [review]
*fawd*

**Comment by nyakase** [comment]
I haven’t actually reviewed the code sorry it’s too long. Im hoping you won’t turn your own thing into a pipe bomb

**Comment by emilianya** [comment]
> I haven’t actually reviewed the code sorry it’s too long. Im hoping you won’t turn your own thing into a pipe bomb

truly "lets gamble, try merging"

**Comment by emilianya** [comment]
This also includes a significant spec change:
image

**Comment by nyakase** [comment]
Second to last paragraph uses the old protocol. This is so siwwy

**Comment by emilianya** [comment]
Updated features.html to match with the deeplink change. For permission system I haven't added anything yet, See: https://youtrack.litdevs.org/issue/LIGHTQUARK-53

**Comment by emilianya** [comment]
> Second to last paragraph uses the old protocol. This is so siwwy

What do you mean?

**Comment by emilianya** [comment]
Ah, the spec
image
Fixed

File: src/routes/v2/auth.ts [hunk 1]
```diff
@@ -6,13 +6,14 @@ import NotFoundReply from "../../classes/reply/NotFoundReply.js";
...
 import * as jose from "jose";
-import UnauthorizedReply from "../../classes/reply/UnauthorizedReply.js";
 import {getNick} from "../../util/getNickname.js";
+import Auth from "../../util/Auth.js";
 
 const router = express.Router();
 const secret = new TextEncoder().encode(process.env.JWT_SECRET);
 
-router.post("/token", (req, res) => {
+// FIXME
```
Comment (AnnoyingRain5) [inline src/routes/v2/auth.ts:16]: You should probably get to this

File: src/routes/v2/auth.ts [hunk 2]
```diff
@@ -32,95 +33,47 @@ router.post("/token", (req, res) => {
...
-                    .setAudience('Lightquark-client')
-                    .setExpirationTime('1y')
-                    .sign(secret).then(jwt => {
-                    return res.json(new Reply(200, true, { message: "Here are your tokens!", token_type: "Bearer", access_token: jwt }))
-                })
-            });
-        })
-    })
+        // User is "logged in" - Password is correct and matches provided email
+        // Lets pull some tokens from ** *** ***
```
Comment (AnnoyingRain5) [inline src/routes/v2/auth.ts:52]: from... where?

File: src/util/ConstantID.ts
Comment (AnnoyingRain5) [inline src/util/ConstantID.ts:1]: why in the world are they in this order

Comment (emilianya) [reply src/util/ConstantID.ts:1]: its the order i added them, but then i went back and sorted them later

File: src/routes/v2/auth.ts (hunk 2 above)
Comment (emilianya) [reply src/routes/v2/auth.ts:52]: i think its "up our ass"

File: src/routes/v2/auth.ts (hunk 1 above)
Comment (emilianya) [reply src/routes/v2/auth.ts:16]: not in scope of #17 (and i dont remember what there is to fix)

**Comment by emilianya** [comment]
@AnnoyingRain5 approval pleas :(

**Pull request merged**